plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -32,25 +36,26 @@
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.collect.MapBuilder;
 import org.elasticsearch.common.unit.TimeValue;
-import org.elasticsearch.common.util.LazyInitializable;
+import org.elasticsearch.common.util.Maps;
 
 import java.io.IOException;
 import java.net.URI;
 import java.util.Map;
-import java.util.concurrent.atomic.AtomicReference;
-import java.util.stream.Collectors;
 
 import static java.util.Collections.emptyMap;
 
 public class GoogleCloudStorageService {
-    
+
     private static final Logger logger = LogManager.getLogger(GoogleCloudStorageService.class);
 
+    private volatile Map<String, GoogleCloudStorageClientSettings> clientSettings = emptyMap();
+
     /**
      * Dictionary of client instances. Client instances are built lazily from the
-     * latest settings.
+     * latest settings. Each repository has its own client instance identified by
+     * a key created using {@link #clientKey(String, String)}.
      */
-    private final AtomicReference<Map<String, LazyInitializable<Storage, IOException>>> clientsCache = new AtomicReference<>(emptyMap());
+    private volatile Map<String, Storage> clientCache = emptyMap();
```

Comment:
The approach here looks fine now in general but we need to enhance it to deal with closing repositories somehow don't we?
Otherwise, we get two problems:

1. If we close a repository its client will still be in this map, effectively creating a memory leak? (looks like we're actually making a similar mistake in `org.elasticsearch.repositories.s3.S3Service#derivedClientSettings` that we could fix eventually ... but leaking full `Storage` instances seems a little scarier since they hold on to non-trivial resources with the internal http client and such)
2. If we close a repository and re-create it at a later date, we'll the stats instance from the previous repository right? (see below)